### DISCUSSION TITLE
Custom Query Parameters in URL

### DISCUSSION THREAD

**Discussion by joshpalmeroc** [algolia/instantsearch#6278] [Q&A] [answered]
I have a search form with two fields "JobPostingTitle" and "Location"

When the user does a search using useSearchBox the "query" query param gets updated in the url and concatenates the two values together as a string. "query=senior%20oregon". This information is fine for algolia to function correctly but i would like to also store additional information in the url like so "query=senior%20oregon&JobPostingTitle=senior&Location=oregon" That way i can pre-populate my form with "JobPostingTitle" or "Location" if someone ever decides to share the url. 

I started to explore routing prop with custom state mapping. My thought was when a user fills out my form i can update the queryParams that i'm holding in useState and pass that on to my stateToRoute function. Although this didn't seem to work as i expected bc stateToRoute function doesn't get access to state changes that happen to queryParams. queryParams in stateToRoute remain unchanged as {} even if i call the setQueryParams function.

```js
const [queryParams, setQueryParams] = useState({});

    return (
        <JobSearchProvider {...props}>
            <InstantSearch
                searchClient={searchClient}
                indexName="example"
                future={{
                    preserveSharedStateOnUnmount: true,
                    cleanUrlOnDispose: false,
                }}
                routing={{
                    router: history({
                        cleanUrlOnDispose: false,
                    }),
                    stateMapping: {
                        stateToRoute(uiState) {
                            const state = uiState['example'];
                            console.log(queryParams); // {JobPostingTitle: 'senior', Location: 'oregon'}
                            return {
                                ...state,
                                ...queryParams,
                            };
                        },
                        routeToState(routeState) {
                            return {
                                ['example']: {
                                    ...routeState,
                                },
                            };
                        },
                    },
                }}
```

**Comment by Haroenv** [comment, marked as answer]
What I recommend is the following:

1. non-InstantSearch parameters should always stay in the URL
2. non-InstantSearch URL updates should be done separately

```js
const routing = {
  stateMapping: {
    stateToRoute(uiState) {
      const existingParameters = new URLSearchParams(window.location.search);

      return {
        ...uiState['example'],
        JobPostingTitle: existingParameters.get('JobPostingTitle'),
        Location: existingParameters.get('Location'),
      };
    },
    routeToState({ JobPostingTitle, Location, ...routeState }) {
      return {
        ['example']: routeState,
      };
    },
  },
};
```

Then later when you update the JobPosting or Location, you can do `history.pushState` with those parameters added or changed.

**Reply by twelve17** [reply]
Hi @Haroenv , thanks for posting this example, I've been looking for guidance on how to do this exact use case.

A few questions:

Is the implied contract as follows?

1. As long as `stateToRoute` handles re-injecting the custom parameters into the route, they should not be clobbered by instantsearch.
2. To update custom parameters, they should be done by updating the history state and doing replace(), and this should not have adverse effects on instantsearch's own routing.

Oh, one more tiny question, should `stateToRoute`  spread `uiState` vs `uiState['example']`? Won't the latter cause loss of other index info? e.g.

```
stateToRoute(uiState) {
      const existingParameters = new URLSearchParams(window.location.search);

      return {
        ...uiState,
        ['example']: {
          ...uiState['example'],
          JobPostingTitle: existingParameters.get('JobPostingTitle'),
          Location: existingParameters.get('Location'),
        }
      };
    },
   ```



Cheers

**Reply by Haroenv** [reply]
my example was written with the assumption that there's only one index, you can indeed also use your code to support multiple indices. Your JobPostingTitle etc should be top-level too, it shouldn't be nested as then it will change place in the url
